### PR TITLE
Use evaluate.py as a submodule.

### PR DESCRIPTION
**Pull request by Eleven1Liu**
1. Description: use libmultilabel as a submodule (starting from `evaluate.py`)
```python 
from LibMultiLabel.libmultilabel.evaluate import evaluate
```

2. Usage: 
```python
>> val_metrics = evaluate(model, dataset_loader, config.monitor_metrics)
>> metric_dict = evaluator.get_metric_dict(use_cache=False, threshold=0.2)  # get results in dictionary format
>> metric_dict['P@5']
>> print(val_metrics) # print cached results
```

### CODE REVIEW

File: libmultilabel/evaluate.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+        to 2D array (shape: number of samples * number of classes).
+        """
+        return np.vstack(self.y_pred)
+
+    def get_metrics(self, use_cache=False):
+        if not use_cache:
+            self.eval(self.y_true, self.y_pred)
+        return self.cache_result
+
+    def __repr__(self):
```
Comment (henryyang42) [inline libmultilabel/evaluate.py:100]: It's not guaranteed that ``__repr__`` will be called before ``eval``

File: libmultilabel/evaluate.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+
+    if save_k_predictions > 0:
+        save_top_k_prediction(model.classes, eval_metric.get_y_pred(), predict_out_path, k=save_k_predictions)
+
+    return metrics
+
+
+class MultiLabelMetrics():
+    def __init__(self, num_classes, monitor_metrics):
+        self.num_classes = num_classes
```
Comment (henryyang42) [inline libmultilabel/evaluate.py:49]: ``num_classes`` is not used anymore

File: libmultilabel/evaluate.py [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+
+import numpy as np
+from sklearn.metrics import f1_score, precision_recall_fscore_support
+from tqdm import tqdm
+
+from .metrics import another_macro_f1, precision_recall_at_ks
+from .utils import Timer, dump_log, save_top_k_prediction
+
+
+def evaluate(model, dataset_loader, num_classes, monitor_metrics, label_key='label', predict_out_path=None, save_k_predictions=0):
```
Comment (henryyang42) [inline libmultilabel/evaluate.py:12]: ``num_classes`` is not used in ``MultiLabelMetrics`` anymore, which can be removed

File: libmultilabel/evaluate.py [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+            y_pred (ndarray): a 2d array with predicted labels (shape: batch_size * number of classes)
+        """
+        self.y_true.append(y_true)
+        self.y_pred.append(y_pred)
+
+    def eval(self, y_true, y_pred, threshold=0.5):
+        """Evaluate precision, recall, micro-f1, macro-f1, and P@k/R@k listed in the monitor_metrics."""
+        y_true = np.vstack(self.y_true)
+        y_pred = np.vstack(self.y_pred)
+        precision, recall, micro_f1, _ = precision_recall_fscore_support(y_true, y_pred > threshold, average='micro')
```
Comment (henryyang42) [inline libmultilabel/evaluate.py:69]: It would be great if we use ``sklearn.metrics.classification_report``. By setting ``output_dict=True`` we can get precision, recall and F1 averaged by ``macro`` and ``micro``

File: libmultilabel/evaluate.py [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+        batch_labels = batch[label_key] # set label_key in os.env or sync naming
+        predict_results = model.predict(batch)
+        batch_label_scores = predict_results['scores']
+
+        batch_labels = batch_labels.cpu().detach().numpy()
+        batch_label_scores = batch_label_scores.cpu().detach().numpy()
+        eval_metric.add_values(batch_labels, batch_label_scores)
+
+    metrics = eval_metric.get_metrics()
+    print(eval_metric)
```
Comment (henryyang42) [inline libmultilabel/evaluate.py:37]: We may print metric outside ``evaluate``.

File: libmultilabel/data_utils.py [hunk 6]
```diff
@@ -111,7 +111,7 @@ def load_or_build_text_dict(config, dataset):
     if config.vocab_file:
         logging.info(f'Load vocab from {config.vocab_file}')
         with open(config.vocab_file, 'r') as fp:
-            vocab_list = [PAD] + [vocab.strip() for vocab in fp.readlines()]
+            vocab_list = [PAD] + [vocab.strip() for vocab in fp.readlines()] # <pad>
```
Comment (cjlin1) [inline libmultilabel/data_utils.py:114]: what's the purpose of this change??

File: libmultilabel/utils.py [hunk 7]
```diff
@@ -81,24 +81,20 @@ def dump_log(config, metrics, split):
         json.dump(result, fp)
 
 
-def dump_top_k_prediction(config, classes, y_pred, k=100):
-    """Dump top k predictions to the predict_out_path. The format of this file is:
+def save_top_k_prediction(class_names, y_pred, predict_out_path, k=100):
```
Comment (cjlin1) [inline libmultilabel/utils.py:84]: better to do save_top_k_predictions

**Review by henryyang42 — APPROVED** [review]

File: main.py [hunk 8]
```diff
@@ -142,10 +141,21 @@ def main():
...
+        test_loader = data_utils.get_dataset_loader(config, datasets['test'], model.word_dict, model.classes, train=False)
+        test_metrics = evaluate(model, test_loader, config.monitor_metrics)
+        metric_dict = test_metrics.get_metric_dict(use_cache=False)
+        dump_log(config=config, metrics=metric_dict, split='test')
+        print(test_metrics)
+        if config.save_k_predictions > 0:
+            if not config.predict_out_path:
+                config.predict_out_path = os.path.join(config.result_dir, config.run_name, 'predictions.txt')
+            save_top_k_predictions(model.classes, test_metrics.get_y_pred(
+            ), config.predict_out_path, config.save_k_predictions)
```
Comment (cjlin1) [inline main.py:155]: better to have pred() in line 154?? having ")" in the beginning of 155 is strange

**Review by cjlin1 — DISMISSED** [review]

**Review by henryyang42 — APPROVED** [review]
